drivers/platform/stm32/stm32_pwm.c

```diff
@@ -0,0 +1,272 @@
+	desc->htimer.Init.Prescaler = 0;
+	desc->htimer.Init.CounterMode = TIM_COUNTERMODE_UP;
+	desc->htimer.Init.Period = period;
+	desc->htimer.Init.ClockDivision = TIM_CLOCKDIVISION_DIV1;
+	desc->htimer.Init.AutoReloadPreload = TIM_AUTORELOAD_PRELOAD_DISABLE;
+	desc->htimer.Init.RepetitionCounter = 0;
+	desc->htimer.State = HAL_TIM_STATE_RESET;
+	if (HAL_TIM_Base_Init(&desc->htimer) != HAL_OK)
+		return -EIO;
+
+	if (HAL_TIM_PWM_Init(&desc->htimer) != HAL_OK)
+		return -EIO;
+
+	/* Store the timer period for calculating PWM duty cycle */
+	desc->timer_period = period;
+	desc->timer_chn = ((stm32_pwm_init_param *)param->extra)->timer_chn;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Initialize the PWM
+ * @param desc - STM32 PWM descriptor
+ * @param param - The structure that contains the PWM init parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+static int32_t _init_pwm(struct stm32_pwm_desc *desc,
+			 const struct pwm_init_param *param)
+{
+	TIM_OC_InitTypeDef sConfigOC = {0};
+	uint32_t pwm_pulse_width;
+	float duty_cycle_percentage;
+	uint32_t period = desc->timer_period;
+	uint32_t timer_chn = ((stm32_pwm_init_param *)(param->extra))->timer_chn;
+
+	/* Calculate the percentage duty cycle */
+	duty_cycle_percentage = ((float)param->duty_cycle_ns / param->period_ns) * 100;
+	pwm_pulse_width = (uint32_t)((period + 1) * duty_cycle_percentage) / (100 - 1);
+
+	sConfigOC.OCMode = TIM_OCMODE_PWM1;
+	sConfigOC.Pulse = pwm_pulse_width;
+	sConfigOC.OCPolarity = TIM_OCPOLARITY_HIGH;
+	sConfigOC.OCFastMode = TIM_OCFAST_DISABLE;
+	if (HAL_TIM_PWM_ConfigChannel(&desc->htimer, &sConfigOC, timer_chn) != HAL_OK)
+		return -EIO;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Initialize the PWM GPIO
+ * @param desc - STM32 PWM descriptor
+ * @param param - The structure that contains the PWM init parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+static int32_t _init_pwm_gpio(struct stm32_pwm_desc *desc,
+			      const struct pwm_init_param *param)
+{
+	int32_t ret;
+	struct stm32_pwm_init_param *sparam = param->extra;
+	struct gpio_init_param *pwm_gpio_params = sparam->gpio;
+
+	ret = gpio_get(&desc->gpio, pwm_gpio_params);
+	if (ret)
+		return ret;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Initialize the PWM
+ * @param desc - PWM descriptor
+ * @param param - The structure that contains the PWM init parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t stm32_pwm_init(struct pwm_desc **desc,
+		       const struct pwm_init_param *param)
+{
+	struct pwm_desc *descriptor;
+	struct stm32_pwm_desc *extra;
+	int32_t ret;
+
+	if (!desc && !param)
+		return -EINVAL;
+
+	descriptor = (struct pwm_desc *)calloc(1, sizeof(*descriptor));
+	if (!descriptor)
+		return -ENOMEM;
+
+	extra = (struct stm32_pwm_desc *)calloc(1, sizeof(*extra));
+	if (!extra) {
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	/* Initialize the PWM timer */
+	ret = _init_pwm_timer(extra, param);
+	if (ret)
```

Comment:
I think `free(extra)` should also be performed in the below error checks via a conditional jump.